docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+        bool TryReadGlobalTargetPointer(string globalName, out TargetPointer value);
+
+        Contracts Contract { get; }
+
+        class Contracts
+        {
+            FieldLayout GetFieldLayout(string typeName, string fieldName);
+            bool TryGetFieldLayout(string typeName, string fieldName, out FieldLayout layout);
+
+            object GetContract(string contractName);
+            bool TryGetContract(string contractName, out object contract);
+
+            // Every contract that is defined has a field here. As an example this document defines a MethodTableContract
+            // If the contract is not supported by the runtime in use, then the implementation of the contract will be the base type which
+            // is defined to throw if it is ever used.
+
+            // Insert list of contracts here
+            MethodTableContract MethodTableContract;
+        }
+    }
+
+    // Types defined by contracts live here
+    namespace ContractDefinitions
+    {
+        class CompositeContract
+        {
+            List<Tuple<string, uint>> Subcontracts;
+        }
+
+        class DataStructureContract
+        {
+            string MethodTableName {get;}
+            List<Tuple<string, FieldLayout>> FieldData;
+        }
+
+        // Insert Algorithmic Contract definitions here
+        class MethodTableContract : Init
+        {
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+            public virtual int BaseSize(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+        }
+    }
+
+    namespace ContractImplementation
+    {
+        // Get contract from the predefined contract database
+        static class PredefinedContracts
+        {
+            public static IContract GetContract(string name, uint version, Target target)
+            {
+                // Do some lookup and allocate an instance of the contract requested
+            }
+        }
+
+        class MethodTableContract_1 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _dynamicTypeIdField;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_1(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _dynamicTypeIdField = _target.Contract.GetFieldLayout("MethodTable", "dynamicTypeId");
+                if (_dynamicTypeIdField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _dynamicTyepIdField)); }
```

Comment:
I've concluded that my C# api that I specced out is weirdly low level, and not really conducive to thinking about the algorithm. I'll update that tomorrow so that there is also higher level model that will probably be used by most contracts in reality. (Whether or not the actual cDAC source code can use a high level or low level model is kind of immaterial. I expect we will need both sorts of apis.)